Driver friendly pool cache

vkgnosis opened the pull request:
Replaces the existing cache. Pool fetching takes a significant amount of
time in both the order book and the driver. This cache hopes to reduce
this by caching recently used tokens automatically when the current
block changes, and keeping previous results cached for a number of
blocks.

Follow ups:
- Change the block number parameter in the pool fetching trait to make it clear whether you get a best effort or exact result.
- Maybe make maintenance trait pass block number around.
- Enable it in the driver.
 
### Test Plan
new unit tests, some manual testing

nlordell commented on e2e/tests/services.rs:
```diff
@@ -131,24 +128,18 @@ impl OrderbookServices {
-            }),
-            current_block_stream,
-        );
+        let pool_fetcher = OrderbookPoolFetcher::new(Box::new(PoolFetcher {
```
Is this maybe missing parameters? Otherwise I'm a little confused.

Also, does the pool fetcher need to be added to the list of `maintiners` in the `ServiceMaintenance` instance :point_down:?

vkgnosis replied:
I didn't finish integrating it fully which is why it's a draft. Just wanted to compare the different cache implementations.

fleupold commented on orderbook/src/orderbook_pool_fetcher.rs:
```diff
@@ -0,0 +1,200 @@
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(cache_miss_block, pairs.iter().copied(), &uncached_pools);
+            mutexed.mark_used(pairs.iter().copied());
```
Could this not be an implementation detail of the get method?

fleupold commented on orderbook/src/orderbook_pool_fetcher.rs:
```diff
@@ -0,0 +1,200 @@
+
+impl Mutexed {
+    fn get(&self, pair: TokenPair, block: Option<u64>) -> Option<&[Pool]> {
+        let block = block.or_else(|| self.cached_most_recently_at_block.get(&pair).copied())?;
```
Maybe we should more explicitly document this "inexact" matching (I'm slightly afraid that someone will in the future request latest block and assume it is actually going to be the latest block). We could even make the "approximate_matching" a cofigurable boolean on the outer Fetcher struct.

Alternatively we could make BlockNumber in the PoolFetching struct optional, None meaning "I don't care what block"

vkgnosis replied:
Added a comment. Agree that proper solution is to make this explicit in the block number enum. Would like to do that in future PR.

fleupold reviewed (commented):
Looks good overall (some comments in line). Looking for feedback from the others what they think about the different approaches

nlordell commented on orderbook/src/orderbook_pool_fetcher.rs:
```diff
@@ -0,0 +1,200 @@
+        let mut cache_misses = HashSet::new();
+        let last_update_block;
+        {
+            let mutexed = self.mutexed.lock().unwrap();
```
Not releasing the lock would not allow simultaneous reads and fetching (in case of multiple clients). My intuition leans towards only holding the lock for reading and updating the cache would perform worse.

vkgnosis replied:
I should have noted that in a comment.
This is intentional and has the trade offs you mention. There is a chance that some work is wasted. The upside is that it guarantees no one has to wait for other fetches. My intuition is that it is unlikely that the same tokens are queried at the same time. And once it has happened the result is cached.
There are more complicated solutions but I would like to have the base version first before attempting those.

nlordell replied:
> My intuition is that it is unlikely that the same tokens are queried at the same time.

Given that the base token pairs should be kept up to date with the `run_maintenance`, I also think, as you mentioned, it is more likely to have cache misses for different tokens than the same one. Also, during these cache misses, other would be stuck waiting for tokens that are already cached.

I think its better to optimize for concurrency here than for worse case.

> There are more complicated solutions but I would like to have the base version first before attempting those.

Very much agree. I would even guess that the gains would be fairly marginal from trying to sync the same fetches (as I suggested :point_up:).

nlordell commented:
> Looking for feedback from the others what they think about the different approaches

I definitely prefer not having separate traits for this (i.e. the approach in this PR). Also, I think the implementation looks very nice (although TBH I didn't review super carefully yet).

vkgnosis commented:
Since this got positive feedback I will update the PR to be in mergable state and address comments.

vkgnosis committed:
Add automatically updating block based pool cache

Replaces the existing cache. Pool fetching takes a significant amount of
time in both the order book and the driver. This cache hopes to reduce
this by caching recently used tokens automatically when the current
block changes, and keeping previous results cached for a number of
blocks.

vkgnosis commented on orderbook/src/database/fees.rs:
```diff
@@ -86,7 +86,7 @@ impl Database {
 #[async_trait::async_trait]
 impl Maintaining for Database {
-    async fn run_maintenance(&self) -> Result<()> {
+    async fn run_maintenance(&self, _new_block: u64) -> Result<()> {
```
Removed it. However if the assumption is that maintenance does *not* correspond to the current block changing, then this cache should not implement Maintainable because it should always be updated when the block changes.

fleupold replied:
I see. Maybe it would make sense rename run_maintenance to on_new_block and do this refactoring codebase wide in a separate PR?

vkgnosis replied:
We don't have to make maintenance run on new blocks, maybe that isn't relevant for other components. I am not sure. Just saying that if it is, then this component shouldn't be part of it. No strong feelings either way.
If we do want to change maintenance then yes I'd do it in a separate PR.

vkgnosis commented:
Finished integrating fully with rest of code and added tests.

fleupold commented on shared/src/arguments.rs:
```diff
@@ -61,6 +64,14 @@ pub struct Arguments {
     pub baseline_sources: Vec<BaselineSources>,
+
+    /// The number of blocks kept in the pool cache.
+    #[structopt(long, env, default_value = "10")]
```
Not sure what the best value is, but maybe 4 blocks (~1min on average) would be a good compromise. If we choose this value too high, we might be quoting outdated prices occasionally.

vkgnosis replied:
We probably want a slightly higher value in the driver so we're sure we still have the pool state when we simulate solver solutions.
On the other hand in the order book as you say this is a bound on how outdated a price can be so it can be lower. Although as long as the token is in the LRU it will have updated pools anyway.

fleupold commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,461 @@
+// When pools are requested we mark all those pools as recently used which potentially evicts other
+// pairs from the pair lru cache. Cache misses are fetched and inserted into the cache.
+// Then when the automatic update runs the next time, we request and cache all recently used pairs.
+// For the order book we only care about the "recent" state of the pools. So we can return any
```
nanonit: I really like the design consideration docs, just for "who uses it how" we could maybe be a bit more generic ("some consumers only care about the recent state"). Otherwise I feel like this comment won't age well with the rest of the codebase changing. But also no big deal

fleupold commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,461 @@
+// by the order book.
+
+impl PoolCache {
+    /// number_of_blocks_to_cache: Previous blocks stay cached until the block is this much older
```
nanonit: I believe older blocks would still be cached if they are queried, but not retained on a new block.

fleupold commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,461 @@
+#[async_trait::async_trait]
+impl PoolFetching for PoolCache {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: BlockNumber) -> Result<Vec<Pool>> {
+        // We assume that if "Latest" is used it is okay to return an earlier cached version even if
```
I assume you don't want to pass `Option<BlockNumber>` because this would make the PR too large? Can we do this right after (could have also been a preparational PR before)?

vkgnosis replied:
Yes I didn't want to include it in this PR. Good idea to do it first. Will make a PR for that now.

fleupold commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,461 @@
+
+    fn insert(&mut self, block: u64, pairs: impl Iterator<Item = TokenPair>, pools: &[Pool]) {
+        for pair in pairs {
+            self.cached_most_recently_at_block.insert(pair, block);
```
What do we mean by "recently" here? It looks like we could cache a pair for an old block (e.g. 25 blocks in the past) and would consider this most recently. Should we check that cached_most_recently_at_block is not larger than block?

Also maybe add a unit test for this case?

vkgnosis replied:
Good point. I think we should have another parameter that decides the maximum age of an entry you can get when requesting "latest". This would be decoupled from the number of blocks cache entries stay alive for.

fleupold replied:
I still think we should only insert `block` here if it's higher than what we already have saved. Otherwise, we could query for a very old block (would override `cached_most_recently_at_block`) and then would cache miss despite having information available in a "recent" block. Not a super important edge case, but feels easy enough to do this check here.

fleupold reviewed (approved):
Looks good to me! I'm a bit worried that translating Block::Latest to some old block can lead to confusion once we forget about this PR, so maybe we can address this immediately after (by replacing with Option<BlockNumber> on the trait level or a custom enum).

fleupold commented:
I was trying to run the benchmark locally but am basically getting NotFound responses for all tokens (on staging the same requests gives a response). Maybe something is not quite right yet?

vkgnosis commented:
> Maybe something is not quite right yet?

Last time I tried it was working. Maybe the node you're using doesn't support tracing so bad token detection fails and prevents the token from being used? Anyway, I'll test it again.

vkgnosis committed:
Add configuration for maximum block age to be considered recent

vkgnosis committed:
Fix cache getting cleared on cache miss

vkgnosis commented:
Results of Felix' benchmark. I first tested on main and then on this PR. Configuration:

* made the gas estimator return a constant 33 gwei
* default settings of the new parameters for this PR
* use our developer staging node
* use our staging base tokens
* disable trace api (no bad token detection)

![image](https://user-images.githubusercontent.com/77051586/120633624-a3e1e500-c46a-11eb-9a84-92a49e3bdb84.png)

We can see that the new cache is faster. But also note that the main feature of this PR which is automatically updating recently used pairs, is not very important for the benchmark.

vkgnosis committed:
remove log message because too much noise

bh2smith commented on shared/src/arguments.rs:
```diff
@@ -61,6 +64,18 @@ pub struct Arguments {
+
+    /// The number of blocks kept in the pool cache.
+    #[structopt(long, env, default_value = "10")]
+    pub pool_cache_blocks: NonZeroU64,
```
TIL: NonZeroU64!

bh2smith commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,524 @@
+// Design:
+// The design of this module is driven by the need to always return pools quickly so that end users
+// going through the api do not have to wait longer than necessary:
+// - The mutex is never held while waiting on an async operation (getting pools from the node).
```
Do you mean it is never blocking or "locked"? Ahhh, you mean "held" like owned elsewhere.

vkgnosis replied:
Technically mutexes can be locked and locks can be held so yeah the language here was a little inaccurate.

bh2smith reviewed (approved):
I like it and look forward to interacting with this interface when we implement `WeightedPoolFetching` (since I imagine we should use it there). Will probably have more comments and questions about convenience of use when that time comes. Nice job here. 

Also, I must say that given the complexity of this change and my current inexperience in comprehending the implications of this change, my approval should only count as 1/2.

nlordell commented on orderbook/src/main.rs:
```diff
@@ -214,11 +214,19 @@ async fn main() {
+            args.shared.pool_cache_lru_size,
+            args.shared.pool_cache_maximum_recent_block_age,
+            Box::new(pool_aggregator),
+            current_block_stream.clone(),
```
Is it an issue that we have separate instances of `current_block_stream` where the "current block" might be different for them because of polling offsets?

For example the `maintenance` block stream could see a new block, but the polling is offset in such a way that the `pool_cache` block stream doesn't see it, and then it will, instead, update for the previous block instead of the next one.

**edit**: I see we are using tokio spmc channels (instead of mpsc) so it is meant for a single producer (the spawned fiber) an multiple receivers, so its not an issue at all.

nlordell commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,524 @@
+#[async_trait::async_trait]
+impl PoolFetching for PoolCache {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: Block) -> Result<Vec<Pool>> {
+        // We assume that if "Latest" is used it is okay to return an earlier cached version even if
+        // it is not exactly latest. This should be reflected in the block enum parameter in the
+        // future.
```
I believe this comment is out of date.

nlordell reviewed (approved):
Looks great! Sorry for taking so long to review.

vkgnosis committed:
Fix comments

fleupold commented on shared/src/pool_cache.rs:
```diff
@@ -0,0 +1,521 @@
+            mutexed.insert(cache_miss_block, cache_misses.into_iter(), &uncached_pools);
+        }
+
+        cache_hits.extend_from_slice(&uncached_pools);
```
As a follow up, maybe we should record a prometheus metric for cache hits and cache misses?

fleupold reviewed (approved):
Looks great! Thanks for the benchmark (baseline seems much faster in your setting than in mine) and refactoring the signature up front.
